✨ Support BYO Public IPv4 Pool when provision infrastructure

mtulio opened the pull request:
**What type of PR is this?**

/kind feature
/kind api-change
/kind documentation

**What this PR does / why we need it**:

Introducing support of PublicIpv4Pool to provision base cluster infrastructure consuming public IPv4 (EIP) from a custom Public IPv4 pool brought to AWS.

A subset of changes of this PR is isolated into those PRs:
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4899
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4900


**Which issue(s) this PR fixes** 

Fixes #4887

**Special notes for your reviewer**:

The changes proposes to create Elastic IPs for each resource and zone claiming public IP address when creating the core infrastructure components and Machines in public subnets. The components supported for core infrastructure are Nat Gateways and Network Load Balancer for API server.

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [ ] adds or updates e2e tests (NA)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Introduce the support of provisioning public IPv4 address consuming from a custom Public IPv4 Pool that is brought to AWS. When the field `PublicIpv4Pool` is set with the pool ID, all the network resources which claims public IPv4, such as Network Load Balancers and NAT Gateways, will be created consuming from the custom pool.
```

mtulio reviewed (commented):
Temporary changes to test locally as it has been addressed in https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4892 . It will be removed once merged/rebased;

mtulio commented:
prevent WIP notifications
/uncc  Ankitasw dlipovetsky

mtulio commented:
/test pull-cluster-api-provider-aws-e2e

mtulio commented:
HI @damdo @nrb  - can I have labels to test in this PR?

damdo commented:
/ok-to-test

mtulio commented:
/test pull-cluster-api-provider-aws-e2e

rvanderp3 commented on pkg/cloud/services/ec2/bastion.go:
```diff
@@ -82,6 +82,7 @@ func (s *Service) ReconcileBastion() error {
 			record.Warnf(s.scope.InfraCluster(), "FailedFetchingBastion", "Failed to fetch default bastion instance: %v", err)
 			return err
 		}
+		// TODO(mtulio): add support of BYO IP for bastion
```
does this still need to be addressed?

mtulio replied:
it seems we can drop this for that version. I can't see the machine field `PublicIP` being used in the bastion workflow.
It seems Bastion will always get public subnets, which has map IP on launch by default on CAPA.
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/4c41955afcb2e2bce07af1754e4c34226c07debe/pkg/cloud/services/ec2/bastion.go#L181

rvanderp3 commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -182,20 +182,7 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte, use
-		// If the subnet does not assign public IPs, set that option in the instance's network interface
-		input.PublicIPOnLaunch = ptr.To(!aws.BoolValue(subnets[0].MapPublicIpOnLaunch))
-	}
+	input.PublicIPOnLaunch = ptr.To(ptr.Deref(scope.AWSMachine.Spec.PublicIP, false))
```
why can't you assign `scope.AWSMachine.Spec.PublicIP` as is?

mtulio replied:
yeah. the intention was to also force a value, but considering it will be used only [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4905/files#diff-4d2ea189a70ea097ffeeb50f78cc676a224386d4fe17b1bec9c83a4731d9c379R539), it will be better to assign as is. Thanks

rvanderp3 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+func (s *Service) createLBAddressesFromPublicIpv4Pool(input *elbv2.CreateLoadBalancerInput, role string) error {
+	// Does not require
```
is this comment complete?

mtulio replied:
No. good catch. Fixed. Thanks

rvanderp3 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,45 @@
+		return errors.Wrapf(err, "failed to allocate EIP from Custom Public IPv4 Pool %v to role: %s", s.scope.VPC().PublicIpv4Pool, role)
+	}
+	if len(eips) != len(input.Subnets) {
+		return fmt.Errorf("allocated EIPs (%d) missmatch with the subnet count (%d)", len(eips), len(input.Subnets))
```
```suggestion
		return fmt.Errorf("allocated EIPs (%d) mismatch with the subnet count (%d)", len(eips), len(input.Subnets))
```

rvanderp3 commented on pkg/cloud/services/elb/loadbalancer.go:
```diff
@@ -378,6 +378,19 @@ func (s *Service) createLB(spec *infrav1.LoadBalancer, lbSpec *infrav1.AWSLoadBa
+		}
+	}
+
+	// Subnets and SubnetMappings are mutual exclusive.
```
```suggestion
	// Subnets and SubnetMappings are mutually exclusive.
```

rvanderp3 commented on api/v1beta2/network_types.go:
```diff
@@ -455,6 +455,27 @@ type VPCSpec struct {
+	PublicIpv4Pool *string `json:"publicIpv4Pool,omitempty"`
+
+	// PublicIpv4PoolFallBackOrder defines the fallback when the Public IPv4 Pool has been exhausted,
+	// no more IPv4 address available in the pool. NOTE: this is not implemented yet.
```
> NOTE: this is not implemented yet.

I believe this is reference in other areas of the code.  I believe only none is supported at the moment, is that correct?

mtulio replied:
Good catch. We can remove this note. This vesion implements the fallback [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4905/files#diff-a3d373b4a7875fa24f1be6f3b2ab92a57fca9d7928388e7d0e1f782a6374706dR61-R87)

mtulio commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -182,20 +182,7 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte, use
-		// If the subnet does not assign public IPs, set that option in the instance's network interface
-		input.PublicIPOnLaunch = ptr.To(!aws.BoolValue(subnets[0].MapPublicIpOnLaunch))
-	}
+	input.PublicIPOnLaunch = scope.AWSMachine.Spec.PublicIP
```
@r4f4 the PR 4905 should be the draft to propose the BYOIP changes.

I have one question for this line/change:
Considering:
- subnet map on launch is added by default on CAPA deployments
- user's config for PublicIP on MachineSpec would have precedence, and may not affect, or depend, if subnet has set the field or not

Is there any specific use case to discover the subnets's map on launch attribute to take the decision of setting the MapPublicIpOnLaunch on instance create, instead of always setting it?

r4f4 replied:
I don't think so.

mtulio commented:
[Downstream OpenShift e2e is passing](https://github.com/openshift/installer/pull/8175#issuecomment-2101480527). The downstream is setting `PublicIpv4Pool` both for infrastructure (VPCSpec) and Machine consuming from an Public IPv4 Pool brought to CI account. Starting tests for regular e2e:

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

mtulio commented:
e2e is mostly passing (eks seems to be flake), added docs and fixed the `verify`:

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

mtulio commented:
This PR is ready for review. We have some spots to improve the unit tests, but I would like to know if this approach is feasile by maitainers to cover the request #4887 .

The e2e had timeouts, e2e-eks pass, trying again with fix in verify.

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

Please take a look?
/cc @damdo @nrb  @r4f4

r4f4 commented on api/v1beta2/awsmachine_webhook.go:
```diff
@@ -334,6 +335,32 @@ func (r *AWSMachine) validateRootVolume() field.ErrorList {
+	if r.Spec.ElasticIPPool == nil {
+		return allErrs
+	}
+	if r.Spec.PublicIP == nil {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec.elasticIpPool"), "publicIp must be set to assign custom public IPv4 pools with elasticIpPool"))
+	}
+	if !*r.Spec.PublicIP {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec.elasticIpPool"), "publicIp must be set to 'true' to assign custom public IPv4 pools with elasticIpPool"))
+	}
```
Isn't this enough?
```suggestion
	if ptr.Deref(r.Spec.PublicIP, false) {
		allErrs = append(allErrs, field.Required(field.NewPath("spec.elasticIpPool"), "publicIp must be set to 'true' to assign custom public IPv4 pools with elasticIpPool"))
	}
```

r4f4 replied:
@mtulio I'm sorry, I have a typo in my suggestion. It should be
```go
if !ptr.Deref(r.Spec.PublicIP, false) {
```

r4f4 commented on api/v1beta2/network_types_test.go:
```diff
@@ -848,3 +848,29 @@ func TestSubnets_HasPublicSubnetWavelength(t *testing.T) {
 	}
 }
+
+func TestPublicIpv4PoolFallbackOrder_String(t *testing.T) {
```
Harmless but I don't see why we need this test.

r4f4 commented on controllers/awsmachine_controller_unit_test.go:
```diff
@@ -670,7 +670,6 @@ func TestAWSMachineReconciler(t *testing.T) {
 				ec2Svc.EXPECT().GetInstanceSecurityGroups(gomock.Any()).Return(map[string][]string{"eid": {}}, nil).Times(1)
 				ec2Svc.EXPECT().GetCoreSecurityGroups(gomock.Any()).Return([]string{}, nil).Times(1)
 				ec2Svc.EXPECT().GetAdditionalSecurityGroupsIDs(gomock.Any()).Return(nil, nil)
-
```
Maybe can drop these empty line deletions in this file?

r4f4 commented on pkg/cloud/scope/machine.go:
```diff
@@ -400,3 +400,14 @@ func (m *MachineScope) SetInterruptible() {
+	if m.AWSMachine == nil {
+		return nil
+	}
+	if m.AWSMachine.Spec.ElasticIPPool == nil {
+		return nil
+	}
```
```suggestion
```

r4f4 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+
+	// Only NLB is supported
+	if *input.Type != string(elbv2.LoadBalancerTypeEnumNetwork) {
+		return fmt.Errorf("custom PublicIpv4Pool is supported only with Network Load Balancer type: %s", *input.Type)
```
Nit: what about
```suggestion
		return fmt.Errorf("custom PublicIpv4Pool not supported with Load Balancer type %s. Use Network Load Balancer instead", *input.Type)
```

r4f4 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+
+	// Custom SubnetMappings should not be defined or overridden by user-defined mapping.
+	if len(input.SubnetMappings) > 0 {
+		return fmt.Errorf("custom PublicIpv4Pool is mutual exclusive with SubnetMappings: %v", input.SubnetMappings)
```
Nit: I don't know if there is any benefit to adding the subnet mappings in the error message. The issue is not with those mappings themselves but with any subnet mappings being specified.
```suggestion
		return fmt.Errorf("custom PublicIpv4Pool is mutually exclusive with SubnetMappings")
```

r4f4 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+
+	eips, err := s.netService.GetOrAllocateAddresses(s.scope.VPC().GetElasticIPPool(), len(input.Subnets), getElasticIPRoleName())
+	if err != nil {
+		return fmt.Errorf("failed to allocate address from Public IPv4 Pool %q to role: %s", *s.scope.VPC().GetPublicIpv4Pool(), getElasticIPRoleName())
```
```suggestion
		return fmt.Errorf("failed to allocate address from Public IPv4 Pool %q to role %s: %w", *s.scope.VPC().GetPublicIpv4Pool(), getElasticIPRoleName(), err)
```

r4f4 commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+		return fmt.Errorf("failed to allocate address from Public IPv4 Pool %q to role: %s", *s.scope.VPC().GetPublicIpv4Pool(), getElasticIPRoleName())
+	}
+	if len(eips) != len(input.Subnets) {
+		return fmt.Errorf("allocated address (%d) from pool %q mismatch with the subnet count (%d)", len(eips), *s.scope.VPC().GetPublicIpv4Pool(), len(input.Subnets))
```
Nit:
```suggestion
		return fmt.Errorf("number of eip addresses (%d) from pool %q must match the subnet count (%d)", len(eips), *s.scope.VPC().GetPublicIpv4Pool(), len(input.Subnets))
```

r4f4 commented on pkg/cloud/services/elb/loadbalancer.go:
```diff
@@ -378,6 +378,22 @@ func (s *Service) createLB(spec *infrav1.LoadBalancer, lbSpec *infrav1.AWSLoadBa
+	// Custom, or BYO, Public IPv4 Pool need to be created prior install, and the Pool ID must be
+	// set in the VpcSpec.ElasticIPPool when the Elastic IP must be consumed from public ip address
+	// from user-provided CIDR blocks.
+	if spec.Scheme != infrav1.ELBSchemeInternetFacing {
```
Shouldn't this be
```suggestion
	if spec.Scheme == infrav1.ELBSchemeInternetFacing {
```

mtulio reviewed (commented):
Addressing the initial review, and ask open questions.

mtulio commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

mtulio commented:
The commits/changes I am performing is improving/adding unit tests, there is not block for review.

@damdo @nrb Would you mind adding it to the review queue, please?

mtulio commented:
/assign @damdo @nrb

mtulio commented:
FWIW I just ran a chaos scenario by fulfilling the custom CIDR pool and run the installation with the fallback to Amazon-pool. I found one better approach to check the fallback, just fixed in the last commit.
The details of the tests and results are available in this comment: https://github.com/openshift/installer/pull/8175#issuecomment-2111229833

mtulio commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
It looks like the e2e failed on teardown:

/test pull-cluster-api-provider-aws-e2e

damdo commented on api/v1beta2/network_types.go:
```diff
@@ -1002,3 +1027,57 @@ func (z ZoneType) String() string {
+	//
+	// +kubebuilder:validation:Enum:=amazon-pool;none
+	// +optional
+	PublicIpv4PoolFallBackOrder *PublicIpv4PoolFallbackOrder `json:"publicIpv4PoolFallbackOrder,omitempty"`
```
What should happen if `PublicIpv4Pool` is nil and `PublicIpv4PoolFallBackOrder` is set? Should we add validation for discarding this case or is it fine?

damdo commented on controllers/awscluster_controller_test.go:
```diff
@@ -1206,7 +1206,7 @@ func mockedCallsForMissingEverything(m *mocks.MockEC2APIMockRecorder, e *mocks.M
 			{
 				Name:   aws.String("tag:sigs.k8s.io/cluster-api-provider-aws/role"),
-				Values: aws.StringSlice([]string{"apiserver"}),
+				Values: aws.StringSlice([]string{"common"}),
```
Are these tests passing with this change?

nrb replied:
I'm honestly not sure if this is a bug or not. I'd have to double check the tags created on a real cluster.

Is this particular change required for the BYOIP functionality? Would we be setting ourselves up to make a breaking API change if we kept the tag value at `apiserver` then change it to `common` later?

jcpowermac replied:
Setting this back to `apiserver` breaks this test. Dropping that commit

nrb replied:
Let's keep the `common` tag, since it's passing tests. I wasn't able to see the NAT gateway using the `apiserver` value anywhere, either.

damdo commented on controllers/awscluster_controller_unit_test.go:
```diff
@@ -247,6 +249,36 @@ func TestAWSClusterReconcileOperations(t *testing.T) {
+					PublicIpv4Pool:              aws.String("ipv4pool-ec2-0123456789abcdef0"),
+					PublicIpv4PoolFallBackOrder: ptr.To(infrav1.PublicIpv4PoolFallbackOrderAmazonPool),
+				}
+				g.Expect(err).To(BeNil())
```
For error output's sake it is better to use `.To(Not(HaveOccurred())`
Same below

damdo commented on controllers/awsmachine_controller.go:
```diff
@@ -522,6 +531,18 @@ func (r *AWSMachineReconciler) reconcileNormal(_ context.Context, machineScope *
+	// BYO Public IPv4 Pool: Allocate and associate an EIP to machines with PublicIP and
+	// cluster-wide Public IPv4 Pool configuration. The EIP must be associated after the instance
+	// is created and transictioned from Pending state.
+	// If the instance already have n public IPv4 address (EIP) associated, it will be released.
```
Typo

damdo commented on docs/book/src/topics/bring-your-own-aws-infrastructure.md:
```diff
@@ -274,3 +274,67 @@ The external system must provide all required fields within the spec of the AWSC
+
+Cluster API also provides a mechanism to allocate Elastic IP from the existing Public IPv4 Pool that you brought to AWS[1].
+
+Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to IPv4 charges introduced by AWS in February 2024[2].
```
```suggestion
Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to buying Public IPs from AWS, also considering the changes in charging for this since February 2024[2].
```

damdo commented on docs/book/src/topics/bring-your-own-aws-infrastructure.md:
```diff
@@ -274,3 +274,67 @@ The external system must provide all required fields within the spec of the AWSC
+- Network Load Balancer for API server
+- Machines
+
+Use `BYO Public IPv4` when you have brought to AWS custom IPv4 CIDR blocks and want to cluster automatically from custom pools instead of Amazon-provided pools.
```
```suggestion
Use `BYO Public IPv4` when you have brought to AWS custom IPv4 CIDR blocks and want the cluster to automatically use IPs from the custom pool instead of Amazon-provided pools.
```

damdo commented on docs/book/src/topics/bring-your-own-aws-infrastructure.md:
```diff
@@ -274,3 +274,67 @@ The external system must provide all required fields within the spec of the AWSC
+
+- BYOIPv4 is limited to AWS to selected regions. See more in [AWS Documentation for Regional availability](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html#byoip-reg-avail)
+- The IPv4 address must be provisioned and advertised to the AWS account before the cluster is installed
+- The public IPv4 addresses is limited to the Network Border Group that the CIDR block have been provisioned, and it must be the same of the cluster will be installed.
```
Not sure what you mean here, we can probably rephrase

damdo commented on docs/book/src/topics/bring-your-own-aws-infrastructure.md:
```diff
@@ -274,3 +274,67 @@ The external system must provide all required fields within the spec of the AWSC
+- BYOIPv4 is limited to AWS to selected regions. See more in [AWS Documentation for Regional availability](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html#byoip-reg-avail)
+- The IPv4 address must be provisioned and advertised to the AWS account before the cluster is installed
+- The public IPv4 addresses is limited to the Network Border Group that the CIDR block have been provisioned, and it must be the same of the cluster will be installed.
+- Only NAT Gateways and Netowork Load Balancer for API server will consume from the IPv4 pool defined in the network scope.
```
```suggestion
- Only NAT Gateways and the Network Load Balancer for API server will consume from the IPv4 pool defined in the network scope.
```

damdo commented on pkg/cloud/services/ec2/eip.go:
```diff
@@ -0,0 +1,54 @@
+	return nil
+}
+
+// ReleaseElasticIP release an specific elastic IP based in the role.
```
```suggestion
// ReleaseElasticIP releases a specific elastic IP based on the instance role.
```

damdo commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -182,19 +182,14 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte, use
-		// If the subnet does not assign public IPs, set that option in the instance's network interface
-		input.PublicIPOnLaunch = ptr.To(!aws.BoolValue(subnets[0].MapPublicIpOnLaunch))
+	// Preserve user-defined PublicIp option.
+	input.PublicIPOnLaunch = scope.AWSMachine.Spec.PublicIP
```
Does this constitute a breaking change with respect to the previous behaviour?

mtulio replied:
Resolving after [lgtm comment](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4905#pullrequestreview-2074052262).

damdo commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	// Only NLB is supported
+	if *input.Type != string(elbv2.LoadBalancerTypeEnumNetwork) {
```
Should we add a check on `input != nil` at the top of this function?

damdo commented on pkg/cloud/services/elb/eip.go:
```diff
@@ -0,0 +1,50 @@
+	return fmt.Sprintf("lb-%s", infrav1.APIServerRoleTagValue)
+}
+
+func (s *Service) allocatedPublicIpv4AddressFromByoIPPool(input *elbv2.CreateLoadBalancerInput) error {
```
I think this should be present tense as we are performing the action with this one
```suggestion
func (s *Service) allocatePublicIpv4AddressFromByoIPPool(input *elbv2.CreateLoadBalancerInput) error {
```

damdo commented on pkg/cloud/services/network/eips.go:
```diff
@@ -156,3 +185,75 @@ func (s *Service) getEIPTagParams(role string) infrav1.BuildParams {
+		return false, nil
+	}
+	if !hasFreeIPs {
+		return false, fmt.Errorf("public IPv4 pool %q does not have free IP addresses: want %d, got %d", *publicIpv4Pool, want, freeIPs)
```
```suggestion
		return false, fmt.Errorf("public IPv4 pool %q does not have enough free IP addresses: want %d, got %d", *publicIpv4Pool, want, freeIPs)
```

damdo commented on pkg/cloud/services/network/eips.go:
```diff
@@ -156,3 +185,75 @@ func (s *Service) getEIPTagParams(role string) infrav1.BuildParams {
+}
+
+// publicIpv4PoolHasFreeIPs check if there are N IPs address available in a Public IPv4 Pool.
+func (s *Service) publicIpv4PoolHasFreeIPs(pool *infrav1.ElasticIPPool, want int64) (bool, error) {
```
This function checks if the pool has _at least_ a "want" number of free IPs.
So I'd probably change `publicIpv4PoolHasFreeIPs` to `publicIpv4PoolHasAtLeastNFreeIPs`

damdo commented on pkg/cloud/services/network/eips.go:
```diff
@@ -156,3 +185,75 @@ func (s *Service) getEIPTagParams(role string) infrav1.BuildParams {
+		return false, fmt.Errorf("failed to describe Public IPv4 Pool %q: %w", *publicIpv4Pool, err)
+	}
+	if len(pools.PublicIpv4Pools) != 1 {
+		return false, fmt.Errorf("unexpected number of Public IPv4 Pools. want 1, got %d", len(pools.PublicIpv4Pools))
```
```suggestion
		return false, fmt.Errorf("unexpected number of configured Public IPv4 Pools. want 1, got %d", len(pools.PublicIpv4Pools))
```

damdo reviewed (commented):
Hey @mtulio thanks a lot for this PR!
I've left some comments, mostly nitpicks, catching typos, and some questions, nothing really wrong with it! :)

mtulio commented:
> Hey @mtulio thanks a lot for this PR! I've left some comments, mostly nitpicks, catching typos, and some questions, nothing really wrong with it! :)

Hello @damdo , thanks a lot for your review, good suggestions. All applied. I am leaving only two comments "unresolved" to further discussion. Let me know your thoughts.

mtulio commented on api/v1beta2/network_types.go:
```diff
@@ -1002,3 +1027,57 @@ func (z ZoneType) String() string {
+	// created in public IPv4 subnets. Every IPv4 address, Elastic IP, will be allocated from the custom
+	// Public IPv4 pool that you brought to AWS, instead of Amazon-provided pool.
+	//
+	// +kubebuilder:validation:MaxLength=30
```
I didn't found anything from AWS documentation, this limitation is based in our development based in couple of pools we are allocating while developing the support on OpenShift. The "pool ID" have all the same length. I will need to take more research to try to find something official :)

nrb replied:
According to the AWS SDK docs (https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2#CreatePublicIpv4PoolOutput), the pool ID is generated by AWS and returned to the user when they create a pool. I can't find much about the length of these IDs either, but if it's generated by AWS, it's possible that their algorithm for making these IDs will change in the future.

That said, I doubt it will change much.

mtulio replied:
@nrb thanks for update. So are we find keeping `MaxLength=30`?

nrb replied:
Yes, let's keep max length at 30.

vr4manta commented:
/lgtm

mtulio commented:
Hi CAPA maintainers, I will be OoO in the next three weeks, please also ping @rvanderp3 to align/address items in your review in this PR. Thanks.

damdo reviewed (commented):
Thanks for addressing feedback @mtulio 
Changes
/lgtm

damdo commented:
Let's get an approver review.
Could any of you PTAL
/assign @richardcase @Ankitasw @dlipovetsky @vincepri @nrb 

Thanks!

nrb commented on api/v1beta2/awsmachine_webhook.go:
```diff
@@ -334,6 +336,29 @@ func (r *AWSMachine) validateRootVolume() field.ErrorList {
+		if !strings.HasPrefix(*eipp.PublicIpv4Pool, awsPublicIpv4PoolPrefix) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec.elasticIpPool.publicIpv4Pool"), r.Spec.ElasticIPPool, fmt.Sprintf("publicIpv4Pool must start with %s.", awsPublicIpv4PoolPrefix)))
+		}
+	}
```
Do we need the `eipp.PublicIpv4Pool == nil && eipp.PublicIpv4PoolFallBackOrder != nil` case here?

mtulio replied:
hey @nrb - the validation `validateNetworkElasticIPPool()` must cover the scenarios bellow, which includes your suggestion. Am I missing something?

| PublicIpv4Pool | PublicIpv4PoolFallBackOrder | covered line |
| -- | -- | -- |
| set (valid) | set | pass L361 |
| set | nil  | err L351 |
| invalid value | set | err L355 |
| nil | set | err L358 |
| nil  | nil | pass L361 |

nrb replied:
Ah, I see what I missed - I read L357 as being nested within the check at L349. Thanks for the clarification.

nrb commented on api/v1beta2/network_types.go:
```diff
@@ -477,6 +483,25 @@ func (v *VPCSpec) IsIPv6Enabled() bool {
 
+// GetElasticIPPool returns the custom Elastic IP Pool configuration when present.
+func (v *VPCSpec) GetElasticIPPool() *ElasticIPPool {
+	if v.ElasticIPPool == nil {
```
Could we just `return v.ElasticIPPool` here, and eliminate the if statement?

nrb commented on api/v1beta2/network_types.go:
```diff
@@ -1002,3 +1027,58 @@ func (z ZoneType) String() string {
+	// resource ID starts with 'ipv4pool-ec2'.
+	//
+	// +kubebuilder:validation:MaxLength=30
+	// +kubebuilder:validation:XValidation:rule="self.startsWith('ipv4pool-ec2-')",message="publicIpv4Pool must start with 'ipv4pool-ec2-'"
```
This validation is present as an XValidation rule and in the webhook; do we need both?

nrb commented on docs/book/src/topics/bring-your-own-aws-infrastructure.md:
```diff
@@ -274,3 +274,69 @@ The external system must provide all required fields within the spec of the AWSC
+Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to buying Public IPs from AWS, also considering the changes in charging for this since February 2024[2].
+
+Supported resources to BYO Public IPv4 Pool (`BYO Public IPv4`):
+- Nat Gateways
```
```suggestion
- NAT Gateways
```

nrb commented on pkg/cloud/services/ec2/instances_test.go:
```diff
@@ -2141,7 +2141,7 @@ func TestCreateInstance(t *testing.T) {
 							MapPublicIpOnLaunch: aws.Bool(true),
 						}},
-					}, nil)
+					}, nil).MaxTimes(2)
```
Why set this max? To fail earlier?

rvanderp3 replied:
I believe the mock was failing due to it being called more than once.  I believe the intent here was address that since it is supposed to be called twice.

mtulio replied:
As Richard said, this test was failing due missing call to `DescribeSubnetsWithContext`, when I set `MaxTimes(2)` it was fixed. Now I am revising those tests and it seems the `DescribeSubnetsWithContext` is duplicated, and this change was recently added. I am not gomock expert, but removing the duplicated call ref resolved those issues (duplicated definition and MaxTimes). See the changes in file [pkg/cloud/services/ec2/instances_test.go](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4905/files#diff-3a3c262a621538f386d0529f1d3089210c0b205e0601318cca26e2804b427185)

jcpowermac commented:
/test pull-cluster-api-provider-aws-test

damdo commented:
/test pull-cluster-api-provider-aws-e2e-eks
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-clusterclass
/test pull-cluster-api-provider-aws-e2e-conformance

mtulio commented:
Thanks @jcpowermac and @rvanderp3 for taking care of this PR addressing the review while I was out.

> /test pull-cluster-api-provider-aws-e2e-eks
> /test pull-cluster-api-provider-aws-e2e
> /test pull-cluster-api-provider-aws-e2e-clusterclass
> /test pull-cluster-api-provider-aws-e2e-conformance

Hey @damdo - apologies I killed the jobs while addressing [this comment](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4905#discussion_r1617656457). Let me trigger again:

/test pull-cluster-api-provider-aws-e2e-eks
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-clusterclass
/test pull-cluster-api-provider-aws-e2e-conformance

mtulio commented:
/test pull-cluster-api-provider-aws-e2e-eks

mtulio commented:
> /test pull-cluster-api-provider-aws-e2e-eks

[Unrelated][job] issue:

[job]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4905/pull-cluster-api-provider-aws-e2e-eks/1799212158454599680

>   STEP: Event details for AWSIAMUserBootstrapper : Resource: AWS::IAM::User, Status: CREATE_FAILED, Reason: Resource handler returned message: "Resource of type 'AWS::IAM::User' with identifier 'bootstrapper.cluster-api-provider-aws.sigs.k8s.io' already exists." (RequestToken: fb13356d-5ac6-1c5f-1b83-6ea7aad87fb4, HandlerErrorCode: AlreadyExists) @ 06/07/24 23:02:31.865

>   STEP: Event details for cluster-api-provider-aws-sigs-k8s-io : Resource: AWS::CloudFormation::Stack, Status: ROLLBACK_IN_PROGRESS, Reason: The following resource(s) failed to create: [AWSIAMManagedPolicyControllers, AWSIAMInstanceProfileControllers, AWSIAMInstanceProfileNodes, AWSIAMInstanceProfileControlPlane, AWSIAMManagedPolicyControllersEKS, AWSIAMUserBootstrapper]. Rollback requested by user. @ 06/07/24 23:02:31.865

Looking the IAM resource name `"Resource of type 'AWS::IAM::User' with identifier 'bootstrapper.cluster-api-provider-aws.sigs.k8s.io' already exists` it seems not to be specific for this job run, it makes me wonder if that job supports parallel executions. cc @damdo @nrb 

I will trigger again with a, hopefully, low traffic period (Friday night =] ).

/test pull-cluster-api-provider-aws-e2e-eks

nrb commented:
Thanks for your diligence and patience with this @mtulio!

/approve

damdo reviewed (commented):
Thanks for the great work here @mtulio & others!
/lgtm
